File: src/install_page/result_row.py
from gi.repository import Adw, Gtk, GLib
from enum import Enum


@Gtk.Template(resource_path="/io/github/flattool/Warehouse/install_page/result_row.ui")
class ResultRow(Adw.ActionRow):
	__gtype_name__ = "ResultRow"
	gtc = Gtk.Template.Child

	version_label = gtc()
	branch_label = gtc()
	add_image = gtc()
	sub_image = gtc()
	selected_image = gtc()
	installed_image = gtc()

	class PackageState(Enum):
		NEW = 0
		SELECTED = 1
		ADDED = 2
		INSTALLED = 3

	def idle_stuff(self):
		self.set_title(GLib.markup_escape_text(self.package.name))
		self.set_subtitle(self.package.app_id)
		self.version_label.set_label(GLib.markup_escape_text(self.package.version))
		self.branch_label.set_label(GLib.markup_escape_text(self.package.branch))
		self.version_label.set_visible(len(self.version_label.get_label()) != 0)
		self.branch_label.set_visible(len(self.branch_label.get_label()) != 0)

	def set_state(self, state):
		if state == self.state:
			return

		self.state = state
		self.add_image.set_visible(False)
		self.sub_image.set_visible(False)
		self.selected_image.set_visible(False)
		self.installed_image.set_visible(False)
		match state:
			case self.PackageState.NEW:
				self.set_sensitive(True)
				self.set_tooltip_text(_("Add Package to Queue"))
				self.add_image.set_visible(True)
			case self.PackageState.SELECTED:
				self.set_sensitive(False)
				self.set_tooltip_text(_("Package has been Added to Queue"))
				self.selected_image.set_visible(True)
			case self.PackageState.ADDED:
				self.set_sensitive(True)
				self.set_tooltip_text(_("Remove Package from Queue"))
				self.sub_image.set_visible(True)
			case self.PackageState.INSTALLED:
				self.set_sensitive(False)
				self.set_tooltip_text(_("This Package is Already Installed"))
				self.installed_image.set_visible(True)

	def __init__(self, package, package_state, origin_list_box, **kwargs):
		super().__init__(**kwargs)

		# Extra Object Creation
		self.state = None
		self.package = package
		self.origin_list_box = origin_list_box

		# Connections

		# Apply
		GLib.idle_add(self.idle_stuff)
		self.set_state(package_state)

File: src/install_page/result_row.ts
import Gtk from "gi://Gtk?version=4.0"
import Adw from "gi://Adw?version=1"

import { Child, from, GClass, Property, WatchProp } from "../gobjectify/gobjectify.js"
import { SearchResult } from "../flatpak.js"

type RowKind = "addable" | "added" | "installed" | "removable"

@GClass({ template: "resource:///io/github/flattool/Warehouse/install_page/result_row.ui" })
export class ResultRow extends from(Adw.ActionRow, {
	result: Property.readonly.gobject(SearchResult),
	kind: Property.readwrite.string("addable").as<RowKind>(),
	_suffix_image: Child<Gtk.Image>(),
}) {
	protected _get_tooltip(): string {
		switch (this.kind) {
			case "addable": return _("Add Package to queue")
			case "added": return _("Package has been added to queue")
			case "installed": return _("Package is already installed")
			case "removable": return _("Remove Package from queue")
		}
	}

	protected _get_sensitivity(): boolean {
		return this.kind === "addable" || this.kind === "removable"
	}

	protected _get_icon(): string {
		switch (this.kind) {
			case "addable": return "warehouse:plus-large-symbolic"
			case "added": return "warehouse:check-plain-symbolic"
			case "installed": return "warehouse:selection-mode-symbolic"
			case "removable": return "warehouse:minus-large-symbolic"
		}
	}

	@WatchProp("kind")
	#on_kind_changed(): void {
		if (this.kind === "installed") {
			this._suffix_image.add_css_class("success")
		} else {
			this._suffix_image.remove_css_class("success")
		}
	}
}
